### PR TITLE
Move wpcom/v2/publicize/connections endpoint to publicize package

### PR DESCRIPTION
**Pull request by manzoorwanijk**
## Proposed changes:

* Move `wpcom/v2/publicize/connections` endpoint to publicize package that can be used in every scenario the publicize module is used
    * WPCOM sites
    * Jetpack sites - both Jetpack and Social plugins

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Testing instructions:

* Goto Jetpack admin or social admin page to verify connections management loads the connections fine.
* Apply this PR to your sandbox and point public API to the sandbox
* Goto [Developer console](https://developer.wordpress.com/docs/api/console/)
* Hit WP REST API - `wpcom/v2/sites/<site_id>/publicize/connections`
* Confirm that the site connections are displayed correctly as per the schema in this PR
* Break a connection by removing Jetpack from the service, e.g. [here for Facebook](https://www.facebook.com/settings/?tab=business_tools).
* Add `?test_connections=1` to the endpoint above
* Verify that the tests are run for connections
* Verify that `status` is set to `'broken'` for the broken connection
* Mark a connection as global in Calypso connections screen
* Verify that global is set to true for backwards compatibility
* Point your Jetpack site to the sandbox
* Goto any page which uses `wp.apiFetch` e.g. Block Editor
* In console, run this
```js
await wp.apiFetch({
    path: '/wpcom/v2/publicize/connections'
});
```
* Confirm that the site connections are displayed correctly as per the schema in this PR

Screenshot 2024-12-13 at 10 27 05 AM

### CODE REVIEW

File: projects/packages/publicize/src/rest-endpoints/class-connections-controller.php
Comment (gmjuhasz) [inline projects/packages/publicize/src/rest-endpoints/class-connections-controller.php]: Should we check here to make sure it's an array?

Comment (manzoorwanijk) [reply projects/packages/publicize/src/rest-endpoints/class-connections-controller.php]: I think the emptiness check below in return statement is sufficient.

**Comment by gmjuhasz** [comment]
This looks good to me, @pablinos what do you think?

File: projects/packages/publicize/src/class-publicize-setup.php
Comment (pablinos) [inline projects/packages/publicize/src/class-publicize-setup.php:55]: We may want to think about this some more. As this file doesn't usually get loaded/run on WPCOM, then maybe this logic shouldn't live here.

File: projects/packages/publicize/actions.php
Comment (pablinos) [inline projects/packages/publicize/actions.php:17]: I realise that we're already doing this, but is this necessary? It should be a very rare occasion that we need directly acces the the filter global like this.

Perhaps Publicize_Assets could be initialized at the point that it's needed? We only need that script data for admin requests for example.

File: projects/packages/publicize/src/rest-endpoints/class-connections-controller.php
Comment (pablinos) [inline projects/packages/publicize/src/rest-endpoints/class-connections-controller.php]: I don't think you're going to like me! But I think we could achieve the same thing as this PR with less change by [moving the existing endpoint](https://github.com/Automattic/jetpack/blob/5e270fb3d34bb3f35e6304392cbdf5d95681b4c1/projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/publicize-connections.php#L4) to the package, make the schema changes, and adding in something like this for Jetpack sites, rather than relying on the Publicize XMLRPC mechanism.

We would have to update the code on WPCOM to require the file from the package rather than the Jetpack plugin, but otherwise it should work.

File: projects/plugins/jetpack/changelog/update-move-wpcom-v2-publicize-connections-to-publicize
Comment (pablinos) [inline projects/plugins/jetpack/changelog/update-move-wpcom-v2-publicize-connections-to-publicize:1]: Do you think this should be more significant for this kind of change, or not because it's backwards compatible?

**Review by pablinos — COMMENTED** [review]
I'm happy to look at it as an alternative, but I wonder about reducing the size of the change further by adapting and moving the current endpoint.

**Comment by pablinos** [comment]
This is pretty untested, but is illustrative of what I'm talking about in this comment https://github.com/Automattic/jetpack/pull/40607#discussion_r1887639941

https://github.com/Automattic/jetpack/pull/40637

We'll still need to do something similar to populate the test results, but we could do a similar change to the other endpoint as a first step. The next would be something like moving the internal methods to a lib like you originally had for the Connections.

File: projects/packages/publicize/src/rest-endpoints/class-connections-controller.php
Comment (manzoorwanijk) [reply projects/packages/publicize/src/rest-endpoints/class-connections-controller.php]: This approach makes the method directly usable in script data to set the connections initial state, unlike the approach in https://github.com/Automattic/jetpack/pull/40637/files#r1887886774

File: projects/packages/publicize/actions.php
Comment (manzoorwanijk) [reply projects/packages/publicize/actions.php:17]: The script data only registers the script, it doesn't enqueue it unless needed.

File: projects/plugins/jetpack/changelog/update-move-wpcom-v2-publicize-connections-to-publicize
Comment (manzoorwanijk) [reply projects/plugins/jetpack/changelog/update-move-wpcom-v2-publicize-connections-to-publicize:1]: Yes, a patch because it's backwards compatible.

**Comment by manzoorwanijk** [comment]
Closing in favor of #40637. Let us avoid going back and forth too many times.

File: projects/packages/publicize/src/class-publicize-setup.php
Comment (manzoorwanijk) [reply projects/packages/publicize/src/class-publicize-setup.php:55]: Yes, we need to improve that logic, but I think it's better to do that as a separate change.

File: projects/packages/publicize/actions.php
Comment (pablinos) [inline projects/packages/publicize/actions.php:10]: I'm happier with this approach as I can see a number of the other packages doing it in the same way, but I wonder if we should restrict this to just WPCOM, similar to the Connection package https://github.com/Automattic/jetpack/blob/38efec6d76d09c00e7d69dea2d2e40ad459cef6d/projects/packages/connection/actions.php#L8

It might again be something we do in a follow-up but this means we're circumventing the Config mechanisms. I'll add another comment in class-publicize-setup.php to explain what I mean

File: projects/packages/publicize/src/class-publicize-setup.php
Comment (pablinos) [inline projects/packages/publicize/src/class-publicize-setup.php:53]: This should really be part of `on_jetpack_feature_publicize_enabled`, which is where we're initialising the other REST API elements, as well as registering the site options, which we'll need if we're bringing those features to WPCOM.

We could move this in there and then call `on_jetpack_feature_publicize_enabled`, but that would register the Connections_Post_Field and REST_Controller on WPCOM, as well as other things that might conflict.

What do you think about calling `pre_initialization` in `on_jetpack_feature_publicize_enabled` and then only using the pattern in actions.php for WPCOM? If we did that, we might want to rename the `pre_initialization` method or something.

**Review by pablinos — APPROVED** [review]
I think we're there and we can address things in follow-ups, but we might want to think about the initialisation a bit more before we merge it.

File: projects/packages/publicize/actions.php
Comment (manzoorwanijk) [reply projects/packages/publicize/actions.php:10]: Yeah, we can use that approach. We can do that in a follow up to unblock other tasks here.

File: projects/packages/publicize/src/class-publicize-setup.php
Comment (manzoorwanijk) [reply projects/packages/publicize/src/class-publicize-setup.php:53]: Let us improve that initialization in a follow up PR

Comment (manzoorwanijk) [reply projects/packages/publicize/src/class-publicize-setup.php:53]: I have created a task for it and added it to our backlog.

**Pull request merged**
